### PR TITLE
Fix the handling of `# fmt: skip` when it's at a colon line.

### LINKED ISSUE #2138: fmt: skip for long line of context managers leads to INTERNAL ERROR: Black produced different code on the second pass of the formatter.
Adding `# fmt: skip` after a long `with ... as ...:` line leads to black raising an internal error.

**To Reproduce** Steps to reproduce the behavior:

1. Take this file `foo.py`:
    ```python
    name = "foo.py"


    def foo():
        with open(name) as f1, open(name) as f2, open(name) as f3, open(name) as f4, open(name) as f5:  # fmt: skip
            print(f1, f2, f3, f4, f5)
    ```
2. Run _Black_ on it with these arguments `black foo.py`
3. See error: `error: cannot format foo.py: INTERNAL ERROR: Black produced different code on the second pass of the formatter.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful: /tmp/blk_lejlacgg.log`
4. The helpful diff (attached)

**Expected behavior** Black formats the source file, leaving the line marked to be skipped alone.

**Environment (please complete the following information):**

- Version: black, version 21.4b0
- OS and Python version: Linux, Python 3.9.4

**Does this bug also happen on master?** [Yes it does](https://black.vercel.app/?version=master&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4AEEAJFdAD2IimZxl1N_WmFm63wsQYxa3iKd5NgD9wTuCPA8tXPZto6Z99Hjctu-G04oaNFwP0prhun9D_WwMRE2qvt9XrGWUmhvMlrPsdHF8rb6YbTjYiezBhLlkOtY8bqbawGuKbY8MsSA1n6NE-6hT0_Pz-gdHZUSmcScrbgrszUMecOfLe-OfEh6y2WRdiiwisWNhecAAAAAjHfk5WJS8PcAAa0BhQIAADpttIKxxGf7AgAAAAAEWVo=)

**Additional context** Add any other context about the problem here.

### LINKED ISSUE #2339: fmt: skip removes a comment-only line before it
**Describe the bug**
Comments behave oddly before `# fmt: skip` lines.

**To Reproduce**

```python
1  # fmt: skip
```
This works as expected: [playground](https://black.vercel.app/?version=main&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ABdAEJdAD2IimZxl1N_Wh88hfShH0DHZZA5Q5v65Sarb329o8bdh1eNCBi1PEY4-H-oxvKzc044Q1qsOBXDx_giyUOM6spXAAAAAJulPCuBSr5kAAFeXgK5bYgftvN9AQAAAAAEWVo=)

```python
# comment
1 # fmt: skip
```
Here the comment is removed: [playground](https://black.vercel.app/?version=main&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ABnAExdAD2IimZxl1N_Wg0-D7--OqqW_cT3fv_7jVgyWfNs4p8A-_XMXk499OalVYdcPx-Mg7NZoOGrqNVCaNgsHCqJlMRpZBzRHyOdABomAwAAEjGgXYIipwsAAWho7r3Lzh-2830BAAAAAARZWg==)

```python
# comment
# comment
1 # fmt: skip
```
Multiple comment lines are also removed: [playground](https://black.vercel.app/?version=main&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4AByAE1dAD2IimZxl1N_Wg0-D7--OqqW_cT3qoxJ_e1a0dd6jGXs5HQs_YMMvlaeXVSpNpgAHY4BN3yrM0AipY8KHnFLePUdLULD7ikB_W1DngvAAAAAAC-f7eCT9J8qAAFpc0NFtV0ftvN9AQAAAAAEWVo=)

<details><summary>Outdated info:</summary>
```python
# comment with
1 # fmt: skip
```
This results in internal error (not equivalent to source) because only a "h" character is left on the comment line that acts as a constant statement: [playground](https://black.vercel.app/?version=main&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ABsAFBdAD2IimZxl1N_Wg0-D7--OqqW_aJd1CNWuEia8Gy8w5OqOeOY_J4QCyqJ2voNYAg5xLwXfn8bNP31cgiRTEgBYJWgMcLgkk02-wXQOGBfve-zAMchIZQPOMaoAAFsbWWMzdoftvN9AQAAAAAEWVo=)

```python
# comment with spaces
1 # fmt: skip
```
This has invalid Python 3 syntax because it produces two constants separated by one space `h spaces`: [playground](https://black.vercel.app/?version=main&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ABzAFZdAD2IimZxl1N_Wg0-D7--OqqW_aJd1CNWq46zIfxxG5eWFn_PhkGjL6RpuWZQx-qLmQcxytzPVX-F3cI3xFlal74yTM5xAnklLBk1vQoKeSIYxHWoInAgAAAAFLr7qgx4SpAAAXJ0ehvnah-2830BAAAAAARZWg==)

</details>

**Expected behavior**
The comments should not affect the second line.

### LINKED ISSUE #2421: Weird behaviour (bug?) of `# fmt: skip`
**Describe the bug**

Dear black developers,

First of all, thank you for this wonderful package!
I wanted to skip one line using `# fmt: skip`. Unfortunately for my case this "skip"-statement seems to be "ignored" by black. I think the following example to reproduce makes it more clear.

**To Reproduce**

Consider the following code:

```python
def somefunction(clusters):
    return bool(x)


if __name__ == "__main__":

    class A:
        foooooobaaar = "foobar"

        def __init__(self):
            if 1:
                if 1:
                    if self.foooooobaaar and not somefunction(clusters=self.foooooobaaar): # fmt: skip
                        pass
```
For example:
1. Take this code snippet
2. Run _Black_ on it with these arguments: `black --diff --color --fast <path_to_snippet>.py`
3. See diff output by black

```bash
--- black_issue.py      2021-08-11 15:37:09 +0000
+++ black_issue.py      2021-08-11 15:37:24.005919 +0000
@@ -8,7 +8,11 @@
         foooooobaaar = "foobar"

         def __init__(self):
             if 1:
                 if 1:
-                    if self.foooooobaaar and not somefunction(clusters=self.foooooobaaar): # fmt: skip
+                    if self.foooooobaaar and not somefunction(
+                        clusters=self.foooooobaaar
+                    ):
+
                         pass
+  # fmt: skip
would reformat black_issue.py
All done! ✨ 🍰 ✨
1 file would be reformatted.
```

**Expected behavior**

I expected no line break for the `clusters=self.foooooobaaar`, because of the `# fmt: skip` statement, and also not to move the `# fmt: skip` statement to the end of the snippet with no indentation.

**Environment (please complete the following information):**

- Version: black, version 21.7b0
- OS and Python version: Python 3.8.3, Ubuntu 20.04.2 LTS

**Does this bug also happen on main?**

I checked the latest master branch and the online (vercel) app. Both result in the same output diff.

**Additional context**

Interestingly omitting the `--fast` option results in an error, black seems to reformat twice?

Thank you very much for your help!

Best, Peter

### LINKED ISSUE #2646: "fmt: skip" on colon-terminated lines either fails or moves the directive
**Describe the bug**

Adding `# fmt: skip` after a colon (e.g. on the same line as `def`, `if`, `with` or `class`) either produces the error "Black produced different code on the second pass of the formatter." or moves the `# fmt: skip` directive below the suite (with a 2-space indent, as well as inserting a blank line between the statement introduction and the suite).

**To Reproduce**

For example, take this code:

```python
def f():  # fmt: skip
    pass
```

And run it with these arguments:

```sh
$ black file.py
```

The resulting error is:

> error: cannot format test.py: INTERNAL ERROR: Black produced different code on the second pass of the formatter.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful: /tmp/blk_cwq0dubx.log

```diff
Mode(target_versions=set(), line_length=88, string_normalization=True, is_pyi=False, is_ipynb=False, magic_trailing_comma=True, experimental_string_processing=False)
--- source
+++ first pass
@@ -1,2 +1,6 @@
-def f(): # fmt: skip
+def f():
+
     pass
+
+
+# fmt: skip
--- first pass
+++ second pass
@@ -1,6 +1,4 @@
 def f():
 
     pass
-
-
-# fmt: skip
+  # fmt: skip
```

On the other hand, with `if` there is no internal error:

```python
if True: # fmt: skip
    pass
```

But it does become:

```python
if True:

    pass
  # fmt: skip
```

Which is not what I expected. `with` and `for` behave like `if`, `class` behaves like `def`.

**Expected behavior**

In all cases I expected black to exit without error and for the `fmt: skip` directive to remain on the same line.

**Environment**

- Black's version:  black, 21.11b2.dev6+g17e42cb (compiled: no)
- OS and Python version:  Debian Bullseye/ Python 3.9.2

### LINKED ISSUE #2680: #fmt: skip results in change in second pass error
**Describe the bug**

When using `#fmt: skip`, if there is no space between the `#` and `f` then black fails to format due to different result in the second pass.

**To Reproduce**

For example, take this code:
```python
def test():
    x = 1 #fmt: skip
```

https://black.vercel.app/?version=main&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ABwAFNdAD2IimZxl1N_WlbvK5V_5PM4Nceap1_OfA9LqGqzctuugz_qJKhKcm0xODvQpn1SGiQJFVPjQaZCfva5pKHycj-2VXHw1ESugA17W0JffVNXXrlGAAC1vzI706u2xQABb3Hpg-HlH7bzfQEAAAAABFla

The resulting error is:

```
INTERNAL ERROR: Black produced different code on the second pass of the formatter.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful: /tmp/blk_9vf9y7f5.log

Mode(target_versions=set(), line_length=88, string_normalization=True, is_pyi=False, is_ipynb=False, magic_trailing_comma=True, experimental_string_processing=False)
--- source
+++ first pass
@@ -1,2 +1,6 @@
 def test():
     x = 1 #fmt: skip
+  # fmt: skip
+
+
+# fmt: skip
--- first pass
+++ second pass
@@ -1,6 +1,3 @@
 def test():
-    x = 1 #fmt: skip
-  # fmt: skip
-
-
-# fmt: skip
+x = 1 #fmt: skip
+  # fmt: skip  # fmt: skip
```

**Expected behavior**

It should format correctly. I assume this is due to the first pass it being an invalid `fmt` specified without the space and so one is added which makes it valid for the second pass.

**Environment**

Black's version: black, 21.12b0 (compiled: no)

### LINKED ISSUE #3126: Blank lines for functions inside blocks & skip
**Describe the bug**
Black currently inserts an empty line before functions when they are defined inside a block.
I don't want that to happen if a function is the first thing defined in a block, so I try to add a skip (see also #3089).
I get an error as a result:
```
INTERNAL ERROR: Black produced different code on the second pass of the formatter.
```
**To Reproduce**
```
if True: # fmt: skip
    def another_func():
        pass
```
Trace:
```
Mode(target_versions=set(), line_length=88, string_normalization=True, is_pyi=False, is_ipynb=False, magic_trailing_comma=True, experimental_string_processing=False, python_cell_magics=set(), preview=False)
--- source
+++ first pass
@@ -1,3 +1,7 @@
-if True: # fmt: skip
+if True:
+
     def another_func():
         pass
+
+
+# fmt: skip
--- first pass
+++ second pass
@@ -1,7 +1,5 @@
 if True:

     def another_func():
         pass
-
-
-# fmt: skip
+  # fmt: skip
```
Also happens if skip is on the next line (i.e. `def another_func(): # fmt: skip`) or both.

**Expected behavior**

Just ignore the line as expected

**Environment**

- Black's version: black, 22.3.0 (compiled: yes)
- OS and Python version: Ubuntu 20.04.4 LTS x86_64; Linux 5.4.0-110-generic; Python 3.9.7

**Additional context**
Would be great if format was changed to not insert a blank line before the function/class definition if it is first in a block.

### PR DESCRIPTION
**Pull request by yilei**
### Description

Fix the handling of `# fmt: skip` when it's at a colon line.

When the Leaf node with `# fmt: skip` is a NEWLINE inside a `suite` Node, the nodes to ignore should be from the siblings of the parent `suite` Node.

There is a also a special case for the ASYNC token, where it expands to the grandparent Node where the ASYNC token is.

This:
- resolves #2646
- resolves #3126 
- resolves #2680
- resolves #2421 
- resolves #2339 
- resolves #2138

### Checklist - did you ...

- [x] Add a CHANGELOG entry if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

**Comment by yilei** [comment]
FWIW, I'm not quite a fan of my change in [`src/black/linegen.py:visit_async_stmt`](https://github.com/psf/black/pull/3148/files#diff-98d222e4d1eed36d13bf2de2b3685fd3fbe65ff82450f30b77bfc11916cee76d), it's somewhat brittle in terms of potential future grammar additions/changes (it's a special case for how `async` grammars are defined in `blib2to3`).

But given `# fmt: skip/on/off` is currently implemented by converting nodes to `STANDALONE_COMMENT`, I couldn't figure out a better way. I would love to be found wrong :)

**Review by ichard26 — APPROVED** [review]
I'd suggest refactoring `generate_ignored_nodes` into a few subfunctions because it's getting pretty hard to follow the logic, but other than that, LGTM. Great work and thank you very much!

**Review by JelleZijlstra — APPROVED** [review]

**Comment by yilei** [comment]
I extracted a `_generate_ignored_nodes_from_fmt_skip` function to handle `# fmt: skip` separately, does this look good? Let me know if you prefer refactoring `_generate_ignored_nodes_from_fmt_skip` further into more separate functions.

**Review by ichard26 — APPROVED** [review]
Thank you so much!

**Comment by ichard26** [comment]
Feels good to close a PR and _six issues at once_ :)

Thank you @yilei again, this feeling is only possible thanks to you! :black_heart:

**Comment by yilei** [comment]
@ichard26 happy to contribute, and thanks for your help too 🖤
